### ISSUE TITLE
Type(String, JSON) has no subscript members

### ISSUE THREAD

**Issue by simonnarang** [SwiftyJSON/SwiftyJSON#584]
I am using the swift 3 branch in Xcode 8 beta 4 and doing:

``` swift
let groupsJson = JSON(data: response2.data!)

   for menuGroup in groupsJson {

      self.menuGroups.append(MenuGroup(name: menuGroup["name"] as! String, image: menuGroup["name"] as! String, guid: menuGroup["guid"] as! String))

   }

```

But I get the error `Type(String, JSON) has no subscript members`
## 

![screen shot 2016-08-03 at 5 02 26 pm](https://cloud.githubusercontent.com/assets/9156815/17381854/2686ee7c-599c-11e6-947c-ed563c2e6588.png)

**Comment by travisday** [comment]
+1

**Comment by ldiqual** [comment]
@simonnarang Thanks for opening an issue! I believe you want to iterate over the `reponse2.data` array by using a `for in`. I think what you need to do is to use `json.arrayValue` or `json.array` and iterate on that. Something like

```
for menuGroup in groupsJson.arrayValue {
    let name = menuGroup["name"] as! String
}
```

For the sake of safety, I would actually add more safeguards:

```
guard let menuGroups = groupsJson.array else {
    fail()
}
for group in menuGroups {
    guard let name = menuGroup["name"].string, let image = menuGroup["image"].string, let guid = menuGroup["guid"].string else {
        continue
    }
    // append
}
```

because `arrayValue` has an implicit fallback on `[]` and `as! String` may lead to crashes, especially with flexible structures like JSON.

Let me know if it solves your issue.

**Comment by jaymutzafi** [comment]
I am having the same error but with a much simpler setup:

```
let json = JSON(data: data!)
print(json)
let pageResults = json["results"]
```

Third line (not in what variation I write the subscript) xcode says Type 'JSON' has no subscript members
although line 1 and 2 do not complain. 

Any idea why?

**Comment by ldiqual** [comment]
@jaymutzafi At first sight your example should definitely work. Can you share a sample project?

**Comment by jaymutzafi** [comment]
@ldiqual Re-added this code and now it suddenly isn't giving this error anymore. But I've added a few lines just to test it and it's producing nil instead of data.

See line 123-126 in the project: https://github.com/jaymutzafi/Movie-Hop-URLS/blob/SwiftyJSON/Movie%20Hop/TMDBService.swift#L123

`let parsedResult: [String:AnyObject]!
                do {
                    parsedResult = try JSONSerialization.jsonObject(with: data!, options: .allowFragments) as! [String:AnyObject]
                } catch {
                    print("Could not parse the data as JSON: '\(data)'")
                    return
                }
                
                let json = JSON(data!)
                print(json)
                let jsonResults = json["results"].array
                print(jsonResults)
                
                var pageResult = [[String:AnyObject]]()
                pageResult = parsedResult["results"] as! [[String:AnyObject]]
`

Basically with the same original data object, JSONSerialization lines work, but SwiftyJ don't.

**Comment by wongzigii** [comment]
@jaymutzafi  

Your issue is because `let json = JSON(data!)` is actually calling this initializer `init(_ object: Any)`.
If you correct it to `let json = JSON(data: data!)`, then you will get it.
I admit this initializer is misleading somehow,  [as part of SwiftyJSON 3.1.4](https://github.com/SwiftyJSON/SwiftyJSON/pull/664/files#diff-b89f704655f1cdfa5eef8141e87c1abaR91),  in this case if you trying to passing a data object to `init(_ object: Any)`, data will be parsed successfully.

**Comment by hassan-dev** [comment]
> I am having the same error but with a much simpler setup:
> 
> ```
> let json = JSON(data)
> print(json)
> let pageResults = json["results"]
> ```
> Third line (no matter in what variation I write the subscript) xcode says Type 'JSON' has no subscript members
> although line 1 and 2 do not complain.
> 
> Any idea why?

it worked for me, thanks
